File: src/chess/game/pieces/Pawn.java
package game.pieces;

// Import Java modules
import java.awt.Color;
import java.util.ArrayList;
import java.util.Collection;
import game.boardUtils.Square;
import game.utils.Constants;
import game.utils.FileC;
import game.utils.RankC;
import game.utils.UtilMethods;

public class Pawn extends Piece {

    private ArrayList<Square> attackers;
    private boolean doubleMovedOnFirstTurn;

    public Pawn(FileC file, RankC rank, Color color) {
        super("Pawn", file,rank,color);
        attackedPieces = new ArrayList<>();
        attackers = new ArrayList<>();
        doubleMovedOnFirstTurn = false;
    }

    public ArrayList<Square> getAttackers() {
        return(attackers);
    }
    public boolean hasDoubleMovedOnFirstTurn() {
        return(this.doubleMovedOnFirstTurn);
    }

    @Override
    public ArrayList<Square> getPossiblePositions(String pos, Collection<Square> squares) {
        ArrayList<Square> possiblePositions = new ArrayList<>();
        FileC f = FileC.valueOf(pos.substring(0, 1));
        RankC r = RankC.valueOf(pos.substring(1));
        if (this.getColor() == Color.BLACK) {
            possiblePositions.add(this.getSquare(squares, f, RankC.valueOf(Integer.toString(r.rank()-1))));
            if (this.getRank() == RankC.SEVEN) {
                possiblePositions.add(this.getSquare(squares, f, RankC.valueOf(Integer.toString(r.rank()-2))));
            }
        } else {
            possiblePositions.add(this.getSquare(squares, f, RankC.valueOf(Integer.toString(r.rank()+1))));
            if (this.getRank() == RankC.TWO) {
                possiblePositions.add(this.getSquare(squares, f, RankC.valueOf(Integer.toString(r.rank()+2))));
            }
        }
        return(possiblePositions);
    }

    @Override
    public boolean validate(FileC f, RankC r, Collection<Square> squares, King king) {
        int[] pos = UtilMethods.getWindowPos(this.getFile(), this.getRank(), Constants.SQUARE_WIDTH, Constants.SQUARE_HEIGHT);
        ArrayList<Square> possiblePositions = getPossiblePositions(UtilMethods.getGamePos(pos[0], pos[1], Constants.SQUARE_WIDTH, Constants.SQUARE_HEIGHT), squares);
        Square endSquare = this.getSquare(squares, f, r);
        if (possiblePositions.size() == 2 && endSquare == possiblePositions.get(1)) {
            this.doubleMovedOnFirstTurn = true;
        }
        if (possiblePositions.contains(endSquare)) {
            Square startSquare = this.getSquare(squares, f, r);
            Piece temp = endSquare.setPiece(startSquare.getPiece());
            if (!(temp instanceof Empty)) {
                endSquare.setPiece(new Empty(endSquare.getRank(), endSquare.getFile(), temp.getColor()));
            } else {
                endSquare.setPiece(temp);
            }
            boolean validated = !king.check(king.getFile(), king.getRank(), squares);
            endSquare.setPiece(startSquare.getPiece());
            startSquare.setPiece(temp);
            return(validated);
        } else {
            return(false);
        }
    }

    public Piece promotion(String promotedPiece) {
        Piece piece;
        switch (promotedPiece.toLowerCase()) {
            case "bishop":
                piece = new Bishop(this.getFile(), this.getRank(), this.getColor());
                break;
            case "knight":
                piece = new Knight(this.getFile(), this.getRank(), this.getColor());
                break;
            case "queen":
                piece = new Queen(this.getFile(), this.getRank(), this.getColor());
                break;
            case "rook":
                piece = new Rook(this.getFile(), this.getRank(), this.getColor());
                break;
            default:
                throw new IllegalArgumentException("Invalid piece type for promotion");
        }
        return(piece);
    }

    public ArrayList<Square> enPassante(Collection<Square> squares) {
        Square previousSquare = this.getSquare(squares, FileC.valueOf(Character.toString(((int) this.getFile().file())-1)), this.getRank());
        Square nextSquare = this.getSquare(squares, FileC.valueOf(Character.toString(((int) this.getFile().file()+1))), this.getRank());
        if (this.getColor() == Color.WHITE) {
            if (this.getRank() == RankC.FIVE) {
                if (previousSquare.getPiece().getClass() == this.getClass()) {
                    if (((Pawn) previousSquare.getPiece()).hasDoubleMovedOnFirstTurn()) {
                        this.setRank(RankC.SIX);
                        this.setFile(previousSquare.getFile());
                        this.attackedPieces.add(previousSquare);
                    }
                } else if (nextSquare.getPiece().getClass() == this.getClass()) {
                    if (((Pawn) nextSquare.getPiece()).hasDoubleMovedOnFirstTurn()) {
                        this.setRank(RankC.SIX);
                        this.setFile(nextSquare.getFile());
                        this.attackedPieces.add(nextSquare);
                    }
                }
            }
        } else if (this.getColor() == Color.BLACK) {
            if (this.getRank() == RankC.FOUR) {
                if (previousSquare.getPiece().getClass() == this.getClass()) {
                    if (((Pawn) previousSquare.getPiece()).hasDoubleMovedOnFirstTurn()) {
                        this.setRank(RankC.THREE);
                        this.setFile(previousSquare.getFile());
                        this.attackedPieces.add(previousSquare);
                    }
                } else if (nextSquare.getPiece().getClass() == this.getClass()) {
                    if (((Pawn) nextSquare.getPiece()).hasDoubleMovedOnFirstTurn()) {
                        this.setRank(RankC.THREE);
                        this.setFile(nextSquare.getFile());
                        this.attackedPieces.add(nextSquare);
                    }
                }
            }
        }
        return(this.attackedPieces);
    }

}

File: bin/chess/game/pieces/pawn.py
from importlib import import_module
from pathlib import Path
from os.path import join
import pygame
from .piece import Piece
from .bishop import Bishop
from .king import King
from .knight import Knight
from .queen import Queen
from .rook import Rook
from chess.CONSTANTS import WHITE, BLACK
from board_utils import Square
from utils.types import Squares, WindowPosition, GamePosition
from utils.functions import get_game_pos, get_window_pos


class Pawn(Piece):

    def __init__(self, file: str, rank: int, color: pygame.Color, min_x: int, max_x: int, min_y: int, max_y: int, square_width: int, square_height: int, win_width: int, win_height: int) -> None:
        pygame.init()
        super().__init__(file, rank, 'Pawn', color)
        self.min_x = min_x
        self.max_x = max_x
        self.min_y = min_y
        self.max_y = max_y
        self.square_width = square_width
        self.square_height = square_height
        self.win_width = win_width
        self.win_height = win_height
        self.x, self.y = self.piece_x, self.piece_y = get_window_pos(
            self.file, self.rank, self.possible_files, self.square_width, self.square_height)
        self.attacked_pieces = []
        self.double_moved_on_first_turn = False

    def validate(self, position: GamePosition, squares: Squares, king: King) -> bool:
        file, rank = position
        if file != self.file:
            if rank == self.rank:
                return False
            elif isinstance(squares[file+str(rank)].piece, Empty):
                squares[file+str(rank)], squares[self.file+str(self.rank)
                                                 ] = squares[self.file+str(self.rank)], squares[file+str(rank)]
                return not king.check(squares=squares, position=position)
            return False
        else:
            if 0 < abs(int(rank)-self.rank) <= 2:
                squares[file+str(rank)], squares[self.file+str(self.rank)
                                                 ] = squares[self.file+str(self.rank)], squares[file+str(rank)]
                print(squares[file+str(rank)])
                print(squares[self.file+str(self.rank)])
                return not king.check(squares=squares, position=position)
            return False

    def move_forward_twice(self, rank: int, file: str, squares: Squares) -> tuple[int, int]:
        if Self.color == WHITE and rank == 2:
            rank += 2
            if squares[file+str(rank)].piece:
                rank -= 2
        elif self.color == BLACK and rank == 7:
            rank -= 2
            if squares[file+str(rank)].piece:
                rank += 2
        return self.get_window_pos(file, rank)

    def promotion(self, promoted_piece: str, images: dict[pygame.Color, dict[str, list]], position: GamePosition):
        promoted_pieces = [Rook, Knight, Bishop, Queen]
        if promoted_piece in promoted_pieces:
            promoted_piece_lower = promoted_piece.lower()
            promoted_piece_module = import_module(
                promoted_piece_lower, Path(__file__).resolve().parent)
            promoted_piece = getattr(promoted_piece_module, promoted_piece)
            return promoted_piece(images[self.color][promoted_piece], self.file, self.rank, self.color, self.min_x, self.max_x, self.min_y, self.max_y, self.square_width, self.square_height, self.win_width, self.win_height)

    def en_passante(self, squares: dict[str, Square]) -> tuple[Piece]:
        previous_square = squares[self.possible_files[self.possible_files.index(
            self.file)-1]+str(self.rank)]
        next_square = squares[self.possible_files[self.possible_files.index(
            self.file)+1]+str(self.rank)]
        if self.color == WHITE:
            if self.rank == 5:
                if previous_square.piece:
                    if isinstance(previous_square.piece, self.__class__):
                        if previous_square.piece.double_moved_on_first_turn:
                            self.rank = 6
                            self.file = previous_square.piece.file
                            self.attacked_pieces.append(previous_square.piece)
                elif next_square.piece:
                    if isinstance(next_square.piece, self.__class__):
                        if next_square.piece.double_moved_on_first_turn:
                            self.rank = 6
                            self.file = next_square.piece.file
                            self.attacked_pieces.append(next_square.piece)
        if self.color == BLACK:
            if self.rank == 4:
                if previous_square.piece:
                    if isinstance(previous_square.piece, self.__class__):
                        if previous_square.piece.double_moved_on_first_turn:
                            self.rank = 3
                            self.file = previous_square.piece.file
                            self.attacked_pieces.append(previous_square.piece)
                elif next_square.piece:
                    if isinstance(next_square.piece, self.__class__):
                        if next_square.piece.double_moved_on_first_turn:
                            self.rank = 3
                            self.file = next_square.piece.file
                            self.attacked_pieces.append(next_square.piece)
        return self.attacked_pieces
